src/common/imagwebp.cpp

```diff
@@ -0,0 +1,231 @@
+        {
+            wxLogError("WebP: WebPDemux failed.");
+        }
+    }
+    return demux;
+}
+
+bool wxWEBPHandler::LoadFile(wxImage *image, wxInputStream& stream, bool verbose, int index)
+{
+    image->Destroy(); // all examples do this, so I do so as well
+    bool ok = false;
+    WebPDemuxerPtr demux = CreateDemuxer(stream, verbose);
+    if (nullptr != demux)
+    {
+        ok = DecodeWebPFrameIntoImage(image, index, demux, verbose);
+    }
+    return ok;
+}
+
+bool wxWEBPHandler::SaveFile(wxImage *image, wxOutputStream& stream, bool)
+{
+    float quality_factor = 90; // if you change this, update the documentation, too
+    if (image->HasOption(wxIMAGE_OPTION_QUALITY))
+    {
+        quality_factor = image->GetOptionInt(wxIMAGE_OPTION_QUALITY);
+    }
+    size_t output_size = 0;
+    uint8_t * output = nullptr;
+    unsigned char * rgb = image->GetData();
+    if (image->HasAlpha())
+    {
+        unsigned char * alpha = image->GetAlpha();
+        int stride = image->GetWidth() * 4; // stride is the "width" of a "line" in bytes
+        std::vector<unsigned char> rgba(stride * image->GetHeight());
+        size_t rgba_index = 0, rgb_index = 0, alpha_index = 0;
+        for (int pixel_counter = 0; pixel_counter < image->GetWidth() * image->GetHeight(); pixel_counter++)
+        {
+            rgba[rgba_index++] = rgb[rgb_index++]; // R
+            rgba[rgba_index++] = rgb[rgb_index++]; // G
+            rgba[rgba_index++] = rgb[rgb_index++]; // B
+            rgba[rgba_index++] = alpha[alpha_index++]; // A
+        }
+        output_size = WebPEncodeRGBA(rgba.data(), image->GetWidth(), image->GetHeight(), stride, quality_factor, &output);
+    }
+    else
+    {
+        int stride = image->GetWidth() * 3; // stride is the "width" of a "line" in bytes
+        output_size = WebPEncodeRGB(rgb, image->GetWidth(), image->GetHeight(), stride, quality_factor, &output);
+    }
+    stream.WriteAll(output, output_size);
+    WebPFree(output);
+    return (output_size > 0 && stream.LastWrite() == output_size);
+}
+
+/*
+TODO: Write tests for animations, then enable this method (remember the header, too).
+
+int wxWEBPHandler::DoGetImageCount(wxInputStream & stream) {
+    int frame_count = -1;
+    WebPDemuxerPtr demux = CreateDemuxer(stream);
+    if (nullptr != demux)
+    {
+        frame_count = WebPDemuxGetI(demux.get(), WEBP_FF_FRAME_COUNT);
+    }
+    return frame_count;
+}
+*/
+
+bool wxWEBPHandler::DoCanRead(wxInputStream& stream)
+{
+    // check header according to https://developers.google.com/speed/webp/docs/riff_container
+    const std::string riff = "RIFF";
+    const std::string webp = "WEBP";
+    const int buffer_size = 12;
+    char buffer[buffer_size];
+    // it's ok to modify the stream position here
+    stream.Read(buffer, buffer_size);
+    if (stream.LastRead() != buffer_size)
+    {
+        return false;
+    }
+    return std::string(buffer, 4) == riff && std::string(&buffer[8], 4) == webp;
```

Comment:
This is probably not a huge optimization but still, why not just do

```suggestion
    return memcmp(buffer, "RIFF", 4) == 0 && memcmp(buffer + 8, "WEBP", 4) == 0;
```

instead of creating these temporary strings?

Reply:
All I can say that I seem to suffer from C++ poisoning and am usually unaware that `memcmp` even exists. 😅